service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/cmo/UploadCMOController.java

```diff
@@ -0,0 +1,147 @@
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+import uk.gov.hmcts.reform.ccd.client.model.AboutToStartOrSubmitCallbackResponse;
+import uk.gov.hmcts.reform.ccd.client.model.CallbackRequest;
+import uk.gov.hmcts.reform.ccd.client.model.CaseDetails;
+import uk.gov.hmcts.reform.fpl.events.cmo.CMOReadyToSealEvent;
+import uk.gov.hmcts.reform.fpl.model.CaseData;
+import uk.gov.hmcts.reform.fpl.model.HearingBooking;
+import uk.gov.hmcts.reform.fpl.model.common.Element;
+import uk.gov.hmcts.reform.fpl.model.common.dynamic.DynamicList;
+import uk.gov.hmcts.reform.fpl.model.order.CaseManagementOrder;
+import uk.gov.hmcts.reform.fpl.service.cmo.UploadCMOService;
+import uk.gov.hmcts.reform.fpl.service.time.Time;
+
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+
+import static uk.gov.hmcts.reform.fpl.model.order.CaseManagementOrder.from;
+import static uk.gov.hmcts.reform.fpl.utils.CaseDetailsHelper.removeTemporaryFields;
+import static uk.gov.hmcts.reform.fpl.utils.ElementUtils.element;
+
+@Api
+@RestController
+@RequestMapping("/callback/upload-cmo")
+@RequiredArgsConstructor(onConstructor_ = {@Autowired})
+public class UploadCMOController {
+    private static final String[] TRANSIENT_FIELDS = {
+        "uploadedCaseManagementOrder", "pastHearingList", "cmoJudgeInfo", "cmoHearingInfo", "numHearings",
+        "singleHearingsWithCMOs", "multiHearingsWithCMOs", "showHearingsSingleTextArea", "showHearingsMultiTextArea"
+    };
+
+    private final Time time;
+    private final UploadCMOService cmoService;
+    private final ObjectMapper mapper;
+    private final ApplicationEventPublisher publisher;
+
+    @PostMapping("/about-to-start")
+    public AboutToStartOrSubmitCallbackResponse handleAboutToStart(@RequestBody CallbackRequest request) {
+        Map<String, Object> data = request.getCaseDetails().getData();
+        CaseData caseData = mapper.convertValue(data, CaseData.class);
+
+        data.putAll(cmoService.getInitialPageData(caseData.getPastHearings(), caseData.getDraftUploadedCMOs()));
+
+        return AboutToStartOrSubmitCallbackResponse.builder()
+            .data(data)
+            .build();
+    }
+
+    @PostMapping("/mid-event")
+    public AboutToStartOrSubmitCallbackResponse handleLabelMidEvent(@RequestBody CallbackRequest request) {
+        Map<String, Object> data = request.getCaseDetails().getData();
+        CaseData caseData = mapper.convertValue(data, CaseData.class);
+
+        // update judge and hearing labels
+        Object dynamicList = caseData.getPastHearingList();
+        List<Element<HearingBooking>> hearings = cmoService.getHearingsWithoutCMO(caseData.getPastHearings(),
+            caseData.getDraftUploadedCMOs());
+        UUID selectedHearing = cmoService.getSelectedHearingId(dynamicList, hearings);
+        data.putAll(cmoService.getJudgeAndHearingDetails(selectedHearing, hearings));
```

Comment:
What do you think about abstracting this inside the service itself? This contain a bit of logic and orchestration which ideally, shouldn't be there.

Reply:
Yeah was thinking that I had left too much in the controller

Reply:
Probably the whole bit (including dynamicList) yes, although dynamicList is less important for me as it is "preparing the response"  I would consider something like that:

```suggestion
data.putAll(cmoService.prepareJudgeAndHearingDetails(dynamicList, caseData.getPastHearings(), caseDatta.getDraftUploadedCMOs))
```
?

Reply:
Sounds good to me